### PR TITLE
Assignment 3

### PR DESCRIPTION
**Pull request by alvaro-dda**
# Assignment 3 Alvaro

### Data Cleaning

- From Assignment 2, I knew the correlation between 'created_at' and 'order_date' as almost 1 so I removed 'created_at'. I also modified the data type of 'order_date'  to float to be able to use it as a predictor.
- As the dataset was already quite high dimensional and had many observations, for computational purposes, I did frequency encoding for the categorical variables.

### Modelling
I mainly compared 4 different generalised linear models:
- A baseline logistic regression with no regularisation: 'log_reg_base'.
- A logistic regression model with l2 regularisation: 'grid_search_l2'.
- A logistic regression model with l2 regularisation and with the f1 metric as the scorer parameter: 'grid_search_f1'.
- A logistic regression model with l2 regularisation and a smaller data set with half of the observations of each type.

The way I chose the threshold was by choosing the value that made the number of false positives the closest from above, to 1% of the total positives. This was done to avoid obtaining all 0 as predictions and to not have a huge number of false positives as we understand that from a business standpoint, the cost of a false positive is high.

**Notes**: 
L1 regularisations weren't used, as the running time of the models was over 10 minutes each, even when trying different solvers.
For all models I set class_weight = 'balanced', to minimise the effect of an unbalanced dataset.

### Results

Model training and the threshold used were done only with the training data (70%). The validation data was used to decide which model gave the best results. Then, the best-performing model was evaluated on the test data.

We observe how all logistic regression models except the one trained on a balanced training set obtained really similar scores. This reinforces how there isn't a highly correlated predictor and the regularisation parameter was chosen to be really close to 0. Also, by balancing our data set we get rid of important information that causes the validation score to be drastically reduced for that model.

Overall, our model would only identify 0.77% of the items a client would actually buy. However, this model minimises the number of false positives, which is 0.6%. The rationale behind this is that it is better to send fewer recommendations but right ones, than sending more but them being wrong.

### Conclusion

Linear models are really weak for this kind of problem and I would suggest using non-linear ones to see if we could improve our results.

### CODE REVIEW

File: src/module_3/assign_3.md [hunk 1]
```diff
@@ -0,0 +1,980 @@
...
+
+
+```python
+#Changing Data Types
+
+df = df.drop(['created_at'], axis=1)
+df['outcome'] = df['outcome'].astype('int8')
+df['order_date'] = pd.to_datetime(df['order_date'])
+conversion_factor = 1e-9
+df['order_date'] = df['order_date'].astype('int64').astype('float64') * conversion_factor
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: What's going on here? It isn't clear what you are trying to do and what the `conversion_factor` is and why is hardcoded to that value. 

Every time you do a non-standard operation, you should make sure it's clear, ideally encapsulating it into a function with docstring. At least add some comments to explain the logic.

File: src/module_3/assign_3.md [hunk 2]
```diff
@@ -0,0 +1,980 @@
...
+
+
+```
+
+
+```python
+#Reordering columns
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: why?!

Comment (guillerbf) [reply src/module_3/assign_3.md]: If you feel you need to do it I'd do it explicitely with the complete list of columns and their order.

File: src/module_3/assign_3.md [hunk 3]
```diff
@@ -0,0 +1,980 @@
...
+
+    3446
+
+
+
+```python
+#Splitting for test/train/validation, keeping the transactions in the same dataset to avoid data leakeage
+
+random.shuffle(transac_5_items)
+
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: `n_transactions = int(len(transac_5_items)` so you can reuse the variable and make the code cleaner

File: src/module_3/assign_3.md [hunk 4]
```diff
@@ -0,0 +1,980 @@
...
+```
+
+    3446
+
+
+
+```python
+#Splitting for test/train/validation, keeping the transactions in the same dataset to avoid data leakeage
+
+random.shuffle(transac_5_items)
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: As we saw in class, this can be problematic since it doesn't account for the potential temporal trend in the data.

File: src/module_3/assign_3.md [hunk 5]
```diff
@@ -0,0 +1,980 @@
...
+train_data_orders =  transac_5_items[:int(len(transac_5_items)*0.7)]
+val_data_orders = transac_5_items[int(len(transac_5_items)*0.7):int(len(transac_5_items)*0.9)]
+test_data_orders = transac_5_items[int(len(transac_5_items)*0.9):]
+```
+
+### Main Models
+
+
+```python
+train_data_x = df[df['order_id'].isin(train_data_orders)].iloc[:,4:]
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: Again, avoid accessing columns via position, is super error prone. Instead define the list of `feature_cols` and the `label_col` explicitly

File: src/module_3/assign_3.md [hunk 6]
```diff
@@ -0,0 +1,980 @@
...
+    plt.tight_layout()
+    plt.show()
+
+
+```
+
+
+```python
+#Base model with no penalties
+log_model_base = LogisticRegression(penalty= None, n_jobs=-1, class_weight='balanced').fit(train_data_x,train_data_y)
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: As we discuss in class, this balancing is "stacking" many points on each positive sample instance (to give them way more weight to correct for the balance). A potential problem is that we are over representing the specific points that happen to be sampled (from the underlying phenomenon) in my dataset.

File: src/module_3/assign_3.md [hunk 7]
```diff
@@ -0,0 +1,980 @@
...
+
+
+```python
+def best_threshold(y_true:pd.core.series.Series, y_prob:np.ndarray)->int:
+    '''
+    Function to obtain the threshold that minimizes false positive to be the closest possible to 1%.
+    '''
+    mean = y_prob.mean()
+    std = y_prob.std()
+    thresholds = np.arange(mean - 3*std, mean + 3*std, std/5)
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: Why? Is it normally distributed? What if your model produces a bimodal distribution with all the probability mass around 0 for negative class and around 1 for positive class?

File: src/module_3/assign_3.md [hunk 8]
```diff
@@ -0,0 +1,980 @@
...
+        y_pred = (y_prob >= threshold).astype(int)
+        
+        false_positives = metrics.confusion_matrix(y_true, y_pred)[0][1]
+
+        if false_positives > y_true.sum()/100:
+            best_threshold = threshold
+
+    print(f"Best Threshold: {best_threshold}")
+
+    return best_threshold
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: You are not selecting the best, but the last one where `false_positives > y_true.sum()/100`

File: src/module_3/assign_3.md [hunk 9]
```diff
@@ -0,0 +1,980 @@
...
+
+grid_search_l2 = GridSearchCV(log_model_reg, param_grid, cv=5, scoring='accuracy',n_jobs=-1).fit(train_data_x,train_data_y)
+
+print(grid_search_l2.best_params_)
+
+y_train_pred_l2 = grid_search_l2.predict(train_data_x)
+y_train_prob_l2 = grid_search_l2.predict_proba(train_data_x)[:,1]
+
+draw_roc_prc(train_data_y, y_train_prob_l2,"With regularization")
+print(metrics.confusion_matrix(train_data_y,y_train_pred_l2))
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: Your model still not learning anything. Would look at `model.coef_`. It's likely the case that the solver isn't converging because you are not scaling/standardising the data. There may be some warning/error being printed potentially.

File: src/module_3/assign_3.md [hunk 10]
```diff
@@ -0,0 +1,980 @@
...
+
+```python
+#New balanced dataset with 50% 0 incomes and 1 outcomes
+train_data = df[df['order_id'].isin(train_data_orders)]
+majority_class = train_data[train_data['outcome'] == 0]
+minority_class = train_data[train_data['outcome'] == 1]
+
+majority_class_undersampled = resample(majority_class, replace=False, n_samples=len(minority_class), random_state=42)
+
+balanced_df = pd.concat([majority_class_undersampled, minority_class])
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: As we discuss in class, balancing the train set doesn't need to be better, we are discarding a lot of negative data until we get both classes balanced. That information may be useful and have prediction power.

I'd evaluate it against unbalance case always, in my experience balancing doesn't tend to yield improvements.

File: src/module_3/assign_3.md (hunk 1 above)
Comment (guillerbf) [reply src/module_3/assign_3.md]: Btw, order_date as any other ID should never be part of the features. It cannot contain any meaningful information to predict future cases. 

What can be done is to extract generic time-related features from it. For instance, `time_of_day`, `day_of_week`, `day_of_month`, `month_of_year`...

File: src/module_3/assign_3.md [hunk 11]
```diff
@@ -0,0 +1,980 @@
...
+```python
+threshold_balanced = best_threshold(balanced_train_data_y, y_train_prob_balanced)
+```
+
+    Best Threshold: 0.4999481480929332
+
+
+
+```python
+#Linear regression model with Reguarization and F1 scoring
```
Comment (guillerbf) [inline src/module_3/assign_3.md]: The process of training is repeated over and over, could and should be extracted on a function. That function will be reused later to write production-ready code.

**Review by guillerbf — COMMENTED** [review]
* Good use of pandas libraries, apply, groupby, etc.
* Good code for plots.
* **Important**. You should not use `created_at` or `order_date` as a feature. First, with random split of orders you can have huge information leakage (hallowing day with everyone buying pumpkins). Second, it doens't contain any predictive power. The day in history isn't informative. Instead, you can extract time-related features that may be useful to capture some temporal patterns.
* The select best threshold function isn't right. If you want to work on FPR of 1%, you can get them from `fpr, tpr, thresholds = roc_curve(y_true, y_pred)`. Just need the min position where `fpr > 0.01` and get the threshold related to it.
* There are few times in the code where you make some arbitrary hard coded decisions `thresholds = np.arange(mean - 3*std, mean + 3*std, std/5)` and `df['order_date'] = df['order_date'].astype('int64').astype('float64') * conversion_factor`. You should always at least explain the reasoning behind it. Typically if they are hardcoded features they should be global variables o parameters in an external config file.
* Accessing columns via position is a pretty bad idea. Super error prone. I encourage you to define the list of features explicitly. It's just a bunch of strings and makes the code way clearer and easier to maintain.
* The model isn't learning at all. The confusion matrix on training data shows that. I think that should be a point to stop and find out why. In this case, I'd remove `order_date` and apply `standardScaler` before the fit cause I suspect the minimisation isn't converging.

File: src/module_3/assign_3.md (hunk 1 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Done! Removed ID Variables

File: src/module_3/assign_3.md (hunk 2 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Agree, changed this as well.

File: src/module_3/assign_3.md (hunk 3 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: I changed my train/test/split so I no longer use the length of transactions.

File: src/module_3/assign_3.md (hunk 4 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Yeah, changed as well!

File: src/module_3/assign_3.md (hunk 5 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Done!

File: src/module_3/assign_3.md (hunk 6 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: I removed this parameter. I get that if our input dataset is going to look as unbalanced as the training one, it doesn't make sense to balance the classes in training.

File: src/module_3/assign_3.md (hunk 7 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: I didn't think about this, good point. Anyways, I change the way I use it to not fall into this trap.

File: src/module_3/assign_3.md (hunk 8 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Yeah, right. I should have used a distance function to the point instead.

File: src/module_3/assign_3.md (hunk 9 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: I didn't get any warnings, but the coefficients were all getting the same values close to 0.

File: src/module_3/assign_3.md (hunk 10 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Perfect, makes sense

File: src/module_3/assign_3.md (hunk 11 above)
Comment (alvaro-dda) [reply src/module_3/assign_3.md]: Implemented a function for this as well!

**Comment by alvaro-dda** [comment]
Hi Guille,
First, I would like to clarify that every PR I do is before our class, which is the reason why I have many mistakes that are later covered in class. Also, for this revision of my code, I haven't looked at your code but I have obviously inspired on it for implementations like the temporal train / test / split, or the pipelines. I have realised that what helps me the most is to do it before class with my own methods, and then once I've seen your implementations, to try and use them without looking at your code.

**Changes in the notebook:**
-  Removed "Created_at" and "Order_ID" variables, as I agree it doesn't make sense to include ID variables as predictors. Instead, I included some numeric variables like month or day of the Week (which didn't improve predictions).
-  Obtain the FPR thresholds like you said, straight from the AUC curve function.
- Changing the way I define features, where now I explicitly define the names of them to avoid the errors you mentioned.
- Implemented the Standard Scaler and now my model seems like is working way better. This surprises me, as I wasn't getting any error on the conversion of the logistic regression (I made sure warnings were on), and from what I have seen logistic regression model shouldn't be sensitive to the scale of the features in terms of prediction quality.
- Implemented the train / val / test split like we did in class by date.
- Implemented a way to check the ROC curves for FPR from 0 to 0.05%, as judging by how unbalanced the dataset is, it wouldn't make sense to go over this threshold.
- Used categorical variables with feature encoding and date features to try to improve the model predictions, which didn't add any predictive power.

**Comment by alvaro-dda** [comment]
Gracias por los comments en el código! Ayudan bastante.

**Comment by guillerbf** [comment]
Si, creo que tu approach es el mejor para sacarle el máximo:
1. Hacer el trabajo por tu cuenta antes de la solución
2. PR con errores.
3. Solucion del profesore + revisión de (2) te da tanto idea general como feedback particular.
4. Adaptar tu solución con lo aprendido.

Buen trabajo Alvaro ;)
